src/node/mod.rs

```diff
@@ -0,0 +1,1241 @@
+        }
+    }
+
+    /// Resume previously paused node.
+    pub fn resume(mut state: PausedState) -> (Self, Receiver<Event>) {
+        let (action_tx, action_rx) = crossbeam_channel::unbounded();
+        let network_rx = state.network_rx.take().expect("PausedState is incomplete");
+        let (user_event_tx, user_event_rx) = crossbeam_channel::unbounded();
+        let (interface_result_tx, interface_result_rx) = crossbeam_channel::bounded(1);
+
+        let timer = Timer::new(action_tx);
+
+        let (stage, core) = Approved::resume(state, timer);
+
+        info!("Resume");
+
+        let node = Self {
+            stage: Stage::Approved(stage),
+            core,
+            action_rx,
+            action_rx_idx: 0,
+            network_rx,
+            network_rx_idx: 0,
+            user_event_tx,
+            interface_result_tx,
+            interface_result_rx,
+        };
+
+        (node, user_event_rx)
+    }
+
+    /// Register the node event channels with the provided [selector](mpmc::Select).
+    pub fn register<'a>(&'a mut self, select: &mut Select<'a>) {
+        // Populate action_rx timeouts
+        #[cfg(feature = "mock_base")]
+        self.core.timer.process_timers();
+
+        self.action_rx_idx = select.recv(&self.action_rx);
+        self.network_rx_idx = select.recv(&self.network_rx);
+    }
+
+    /// Processes events received externally from one of the channels.
+    /// For this function to work properly, the node event channels need to
+    /// be registered by calling [`ApprovedPeer::register`](#method.register).
+    /// [`Select::ready`] needs to be called to get `op_index`, the event channel index.
+    ///
+    /// This function is non-blocking.
+    ///
+    /// Errors are permanent failures due to either: node termination, the permanent closing of one
+    /// of the event channels, or an invalid (unknown) channel index.
+    ///
+    /// [`Select::ready`]: https://docs.rs/crossbeam-channel/0.3/crossbeam_channel/struct.Select.html#method.ready
+    ///
+    /// The returned `bool` can be safely ignored by the consumers of this crate. It is for
+    /// internal uses only and will always be `true` unless compiled with `feature=mock_base`.
+    pub fn handle_selected_operation(&mut self, op_index: usize) -> Result<bool, RecvError> {
+        if !self.is_running() {
+            return Err(RecvError);
+        }
+
+        match op_index {
+            idx if idx == self.network_rx_idx => {
+                let event = self.network_rx.recv()?;
+                self.handle_network_event(event);
+                Ok(true)
+            }
+            idx if idx == self.action_rx_idx => {
+                let action = self.action_rx.recv()?;
+
+                let status = is_busy(&action);
+                self.handle_action(action);
+                Ok(status)
+            }
+            _idx => Err(RecvError),
+        }
+    }
+
+    /// Returns whether this node is running or has been terminated.
```

Comment:
`Node` can never be paused, because calling `pause` consumes `self` and returns `PausedState`.

Reply:
That's tricky, because the decision whether to terminate depends on the input events and the handlers from those take `&mut self`, not `self`. 

What I'm thinking instead is that we might be able to remove the `Terminated` stage completely. Currently we only use it when we fail to bootstrap. So maybe we can just remain in the `Bootstrapping` stage instead and only raise the `Terminated` event. Functionality-wise should be the same, as both require the user to deal with the situation. Maybe something for a future PR.